Fixed #32243 -- Added docs examples for manually saving Files.

massover opened the pull request:
[Ticket #32243](https://code.djangoproject.com/ticket/32243#comment:7)

After suggesting a solution on the [forums](https://forum.djangoproject.com/t/cant-upload-image-though-manage-py-shell/11231/3), I thought that the documentation could use an example as well. As far as I could find, there wasn't an example for how to upload a file outside of a request context. I know I have done this in tests, celery tasks, and management commands.

apollo13 commented:
I think it would be better to use `django.core.files.base.File/ContentFile` -- I cannot think of a good reason to use stuff from the `uploadedfile` or do I miss something.

massover commented:
> I think it would be better to use `django.core.files.base.File/ContentFile` -- I cannot think of a good reason to use stuff from the `uploadedfile` or do I miss something.

Makes sense. I updated the docs and I added as a section in `Using files in models` as well.

smithdc1 commented on docs/ref/files/uploads.txt:
```diff
@@ -99,6 +99,11 @@ Subclasses of ``UploadedFile`` include:
 
+.. class:: SimpleUploadedFile
+
+    A simple representation of a file, which just has content, size, and a name.
```
We try to avoid words such as just / simple. Although given that's the class name I'm not sure how much we can do here. 

https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#writing-style

massover replied:
Thanks. I removed this documentation change from the PR. After @apollo13's suggestion there's no more reference to `SimpleUploadedFile` with my change.

For more context, the exact copy of the documentation that I used

>  A simple representation of a file, which just has content, size, and a name.

comes directly from the already existing [SimpleUploadedFile](https://github.com/django/django/blob/main/django/core/files/uploadedfile.py#L102)'s docstring. I don't think it needs to be changed. But if you or someone wants to suggest different copy I'm glad to change it in this PR.

kezabelle commented:
Thanks for your work so far.

Personally, I can can on two hands the number of times I've needed to put raw bytes (that I've read from _somewhere_) into a file and then into a model, but I've got _plenty_ of history of having to smash actual files on disk in, which I _imagine_ is the more common requirement.

I'm not objecting to the `ContentFile(b'...')` addition -- I welcome it (the more the merrier given the somewhat _magical_ appearance of uploaded-file-to-storage), but I want to make sure that we're actually covering the use case requested in the forum thread in the most idiomatic way too - that appeared to be "I have a list of files in a directory and I want to save each of them to a model instance [... and persist them to media storage]", paraphrasing fairly heavily. Is that case best served by `ContentFile`, or a `File`?

massover commented:
> Thanks for your work so far.
> 
> Personally, I can can on two hands the number of times I've needed to put raw bytes (that I've read from _somewhere_) into a file and then into a model, but I've got _plenty_ of history of having to smash actual files on disk in, which I _imagine_ is the more common requirement.
> 
> I'm not objecting to the `ContentFile(b'...')` addition -- I welcome it (the more the merrier given the somewhat _magical_ appearance of uploaded-file-to-storage), but I want to make sure that we're actually covering the use case requested in the forum thread in the most idiomatic way too - that appeared to be "I have a list of files in a directory and I want to save each of them to a model instance [... and persist them to media storage]", paraphrasing fairly heavily. Is that case best served by `ContentFile`, or a `File`?

In that case I would use a File. Fwiw before this PR I would have always used SimpleUploadedFile!

Most of my cases are turning bytes into files, dynamically generating pdfs, fetching images from other sources, dynamically generating images, generating static json files.

The important part to me is that we are documenting what I think is a happy path:

- instantiate a File like object
- assign it to a file field
- call save on the instance

I don’t mind whether it’s a File object or ContentFile object in the example. I just chose ContentFile because it’s easier to document and tried to reference “File like” with links to the File object.

claudep commented:
I also think that putting an existing file on disk in a model field should be documented in this PR.

massover commented:
> I also think that putting an existing file on disk in a model field should be documented in this PR.

@claudep / @kezabelle 

There are now 2 examples. The example in `Using files in models` includes an example saving an existing image file from disk to a file field. The `File Uploads` section contains an example outside of a request using a command with a content file.

claudep commented on docs/topics/files.txt:
```diff
@@ -94,6 +94,17 @@ location (:setting:`MEDIA_ROOT` if you are using the default
+    >>> f = open(settings.MEDIA_ROOT + 'cars/chevy_iii.jpg', 'rb')
+    >>> car = Car.objects.get(name='57 Chevy')
+    >>> car.photo = ImageFile(f)
+    >>> car.save()
```
Thanks for this addition. You should use a `with` construct to avoid letting the file unclosed (unless the save() closes the file somehow, I didn't test).

However, this is suitable when a file has already been uploaded inside media root. I think a more common use case is when the file comes as an external source. Here's the code I generally use:

```
from pathlib import Path
from django.core.files import File

file_path = Path('/some/external/file.pdf')
car = Car.objects.get(name='57 Chevy')
with file_path.open(mode='rb') as handle:
    car.specs = File(handle, name=file_path.name)
    car.save()
```

massover replied:
excellent suggestion and example, thank you! updated the example.

massover commented:
@carltongibson anything else you think we can improve?

felixxm commented on docs/topics/files.txt:
```diff
@@ -94,6 +95,16 @@ location (:setting:`MEDIA_ROOT` if you are using the default
+    >>> path = Path('/some/external/specs.pdf')
+    >>> car = Car.objects.get(name='57 Chevy')
+    >>> with path.open(mode='rb') as f:
+    >>>     car.specs = File(f, name=path.name)
+    >>>     car.save()
```
```suggestion
    ...     car.specs = File(f, name=path.name)
    ...     car.save()
```

massover replied:
Fixed! Thanks

felixxm commented on docs/topics/files.txt:
```diff
@@ -94,6 +95,16 @@ location (:setting:`MEDIA_ROOT` if you are using the default
         >>> image
         <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=191x287 at 0x7F99A94E9048>
 
+To save an existing file on disk to a :class:`~django.db.models.FileField`::
+
+    >>> from pathlib import Path
+    >>> from django.core.files import File
+    >>> path = Path('/some/external/specs.pdf')
+    >>> car = Car.objects.get(name='57 Chevy')
+    >>> with path.open(mode='rb') as f:
+    >>>     car.specs = File(f, name=path.name)
+    >>>     car.save()
```
This should be moved above the note.

massover replied:
Fixed! Thanks

felixxm commented on docs/topics/http/file-uploads.txt:
```diff
@@ -126,6 +126,19 @@ model::
         return render(request, 'upload.html', {'form': form})
 
+If you are constructing an object manually outside of a request, you can assign
+the :class:`~django.db.models.FileField` to a :class:`~django.core.files.File`
+like object on the model instance::
```
As far as I'm aware `FileField` and `File` should be swapped in this sentence
```suggestion
a :class:`~django.core.files.File` like object to the
:class:`~django.db.models.FileField`::
```

massover replied:
Makes sense. Fixed!

carltongibson reviewed (approved):
Hey @massover — thanks for this! 

I got knocked somewhat sideways by the start of 2022, so missed the discussion entirely, but it looks like all the key issues have been addressed, and both the new examples look super. 

I'll mark this _Ready for Checkin_ and merge in the morning if no-one shouts out. 

Welcome aboard! ⛵

carltongibson commented:
Im going to call this `Fixed #32243 -- ...` — I don't see too much left there with these examples in place.

carltongibson commented:
@massover I rebased and adjust the commit message format. Thanks again!

felixxm commented on docs/topics/http/file-uploads.txt:
```diff
@@ -126,6 +126,18 @@ model::
         return render(request, 'upload.html', {'form': form})
 
+If you are constructing an object manually outside of a request, you can assign
+the :class:`~django.core.files.File` like object to the :class:`~django.db.models.FileField`::
```
Wrap at 79 chars:
```suggestion
the :class:`~django.core.files.File` like object to the
:class:`~django.db.models.FileField`::
```
